otx/api/usecases/reporting/time_monitor_callback.py

```diff
@@ -67,6 +72,29 @@ def __init__(
 
         self.update_progress_callback = update_progress_callback
 
+    def __getstate__(self):
+        """Return state values to be pickled."""
+        state = self.__dict__.copy()
+        # update_progress_callback is not always pickable object
+        # if it is not, replace it with default callback
+        if not dill.pickles(state["update_progress_callback"]):
+            state["update_progress_callback"] = default_progress_callback
+        return state
+
+    def __deepcopy__(self, memo):
+        """Return deepcopy object."""
+        result = self.__class__(
+            self.total_epochs,
+            self.train_steps,
+            self.val_steps,
+            self.test_steps,
+            self.epoch_history,
+            self.step_history,
+            self.update_progress_callback,
+        )
+        memo[id(self)] = result
+        return result
```

Comment:
Could you let me know why this is needed?
(And the line#95)
Thanks!

Reply:
`__deepcopy__` is a special method for `copy.deepcopy` and unless it is implemented, `copy.deepcopy` falls back to use `__[gs]etstate__` indirectly. Since we implemented `__getstate__`, `copy.deepcopy` will use `__getstate__`, which means the deep copied object could not be a homogeneous object as we replaced the unpickable `update_progress_callback` object with sensible default one `default_progress_callback`. To avoid this, we should implement `__deepcopy__` method.

`memo` is to prevent recursive infinite loop if the object is recursive. Please refer [here](https://docs.python.org/3/library/copy.html#copy.deepcopy).

Reply:
Thank you for kind explanation. Understood clearly!